File: tmdb-core/src/main/java/dev/reuss/tmdb/domain/search/query/SearchMultiQuery.java
package dev.reuss.tmdb.domain.search.query;

import dev.reuss.tmdb.core.http.QueryParams;
import dev.reuss.tmdb.domain.search.SearchQuery;
import dev.reuss.tmdb.query.PagedQuery;
import dev.reuss.tmdb.query.QueryValidation;
import dev.reuss.tmdb.value.language.Language;

/**
 * Query parameters for searching movies, TV shows and people.
 *
 * <p>The search text is required and must not be blank. Optional filters are
 * adult content inclusion, language and page. Page values are validated by
 * {@link QueryValidation#validatePage(Integer)}.</p>
 */
public final class SearchMultiQuery implements SearchQuery, PagedQuery<SearchMultiQuery> {

    private final String query;
    private Boolean includeAdult;
    private Language language;
    private Integer page;

    private SearchMultiQuery(String query) {
        this.query = QueryValidation.requireNotBlank(query, "Search query");
    }

    public static SearchMultiQuery of(String query) {
        return new SearchMultiQuery(query);
    }

    @Override
    public String query() {
        return query;
    }

    public SearchMultiQuery includeAdult(Boolean includeAdult) {
        this.includeAdult = includeAdult;
        return this;
    }

    public SearchMultiQuery language(Language language) {
        this.language = language;
        return this;
    }

    public SearchMultiQuery page(Integer page) {
        QueryValidation.validatePage(page);
        this.page = page;
        return this;
    }

    @Override
    public QueryParams toQueryParams() {
        return QueryParams.create()
                .add("query", query)
                .add("include_adult", includeAdult)
                .add("language", language)
                .add("page", page);
    }
}

File: tmdb-core/src/main/kotlin/dev/reuss/tmdb/domain/search/query/SearchMultiQuery.kt
package dev.reuss.tmdb.domain.search.query

import dev.reuss.tmdb.core.http.QueryParams
import dev.reuss.tmdb.query.PagedQuery
import dev.reuss.tmdb.query.QueryValidation
import dev.reuss.tmdb.value.language.Language

/**
 * Query parameters for searching movies, TV series, and people in one request.
 *
 * The search text is required and must not be blank. Optional filters are
 * adult content inclusion, language and page. Page values are validated by
 * [QueryValidation.validatePage].
 */
class SearchMultiQuery private constructor(
    override val query: String,
) : SearchQuery,
    PagedQuery<SearchMultiQuery> {
    private var includeAdult: Boolean? = null
    private var language: Language? = null
    private var page: Int? = null

    /** Controls whether adult results may be returned; `null` uses the TMDB default. */
    fun includeAdult(value: Boolean?) =
        apply {
            includeAdult = value
        }

    /** Sets the response language; `null` leaves the parameter unspecified. */
    fun language(value: Language?) =
        apply {
            language = value
        }

    /** Sets the one-based result page; `null` leaves the parameter unspecified. */
    override fun page(page: Int?) =
        apply {
            QueryValidation.validatePage(page)
            this.page = page
        }

    override fun toQueryParams(): QueryParams =
        QueryParams
            .create()
            .add("query", query)
            .add("include_adult", includeAdult)
            .add("language", language?.value)
            .add("page", page)

    companion object {
        /**
         * Creates a query for trimmed [query] text.
         *
         * @throws IllegalArgumentException if [query] is blank
         */
        @JvmStatic
        fun of(query: String): SearchMultiQuery =
            SearchMultiQuery(
                QueryValidation.requireNotBlank(query, "Search query"),
            )
    }
}
